### PR TITLE
SI-8447 fix TypeTree printing

### PR DESCRIPTION
**Pull request by VladimirNik**
Added TypeTree printing based on tpe to CodePrinter (if original is not available) - [SI-8447](https://issues.scala-lang.org/browse/SI-8447)

### CODE REVIEW

**Comment by VladimirNik** [comment]
Review by @xeno-by @densh

File: test/junit/scala/reflect/internal/PrintersTest.scala [hunk 1]
```diff
@@ -715,7 +727,7 @@ trait ClassPrintTests {
     |}""",
     typedCode = sm"""
     |object PM5 {
-    |  scala.collection.immutable.List.apply(1, 2) match {
+    |  scala.collection.immutable.List.apply[Int](1, 2) match {
```
Comment (xeno-by) [inline test/junit/scala/reflect/internal/PrintersTest.scala:730]: Why's this printing an inferred type argument? I was under impression that `wasEmpty` handles this.

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 2]
```diff
@@ -599,6 +599,8 @@ trait Trees extends api.Trees {
       */
     private[scala] var wasEmpty: Boolean = false
 
+    def wasEmptyValue = wasEmpty
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Trees.scala]: Isn't `wasEmpty` available to the printer anyway?

File: src/reflect/scala/reflect/internal/Printers.scala [hunk 3]
```diff
@@ -1061,7 +1070,11 @@ trait Printers extends api.Printers { self: SymbolTable =>
           print("(", qualifier, ")#", blankForOperatorName(selector), printedName(selector))
 
         case tt: TypeTree =>
-          if (!emptyTree(tt)) print(tt.original)   
+          if (!emptyTree(tt)) {
+            val original = tt.original
+            if (original != null) print(original)
+            else super.printTree(tree)
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Printers.scala:1076]: Can we check for wasEmpty here?

File: src/reflect/scala/reflect/internal/Printers.scala [hunk 4]
```diff
@@ -984,8 +993,8 @@ trait Printers extends api.Printers { self: SymbolTable =>
             case treeInfo.Unapplied(body) =>
               body match {
                 case Select(qual, name) if name == nme.unapply  => print(qual)
-                case TypeApply(Select(qual, name), args) if name == nme.unapply || name == nme.unapplySeq =>
-                  print(TypeApply(qual, args))
+                case TypeApply(Select(qual, name), _) if name == nme.unapply || name == nme.unapplySeq =>
+                  print(qual)
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Printers.scala:997]: Why this change?

File: src/reflect/scala/reflect/internal/Printers.scala [hunk 5]
```diff
@@ -596,18 +596,26 @@ trait Printers extends api.Printers { self: SymbolTable =>
...
-      case EmptyTree | build.SyntacticEmptyTypeTree() => true
-      case _ => false
+    object EmptyTypeTree {
+      def unapply(tt: TypeTree): Boolean = tt match {
+        case build.SyntacticEmptyTypeTree() if tt.wasEmptyValue || tt.isEmpty => true
+        case _ => false
+      }
     }
 
+    protected def emptyTree(tree: Tree) =
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Printers.scala]: Probably the `is` prefix would be appropriate here

File: src/reflect/scala/reflect/internal/Printers.scala (hunk 4 above)
Comment (VladimirNik) [inline src/reflect/scala/reflect/internal/Printers.scala:997]: If we omit unapply function, we also should remove its type parameters. In earlier version they are not printed because we don't print TypeTree without original.

File: test/junit/scala/reflect/internal/PrintersTest.scala (hunk 1 above)
Comment (VladimirNik) [inline test/junit/scala/reflect/internal/PrintersTest.scala:730]: Yes, it was surprise for me too. Here (wasEmpty || isEmpty) is false

File: src/reflect/scala/reflect/internal/Printers.scala (hunk 4 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Printers.scala:997]: What if the type arguments were explicitly specified by the user?

File: test/junit/scala/reflect/internal/PrintersTest.scala (hunk 1 above)
Comment (xeno-by) [inline test/junit/scala/reflect/internal/PrintersTest.scala:730]: Wow. So what's the original?

File: src/reflect/scala/reflect/internal/Printers.scala (hunk 4 above)
Comment (VladimirNik) [inline src/reflect/scala/reflect/internal/Printers.scala:997]: How do we specify type if we have code without manually written invocation of unapply, like:
val a(x, y) = b
or case a(x, y) => ...

Could you please write an example with explicit type arguments when outer tree is UnApply and its fun is treeInfo.Unapplied ?

File: test/junit/scala/reflect/internal/PrintersTest.scala (hunk 1 above)
Comment (VladimirNik) [inline test/junit/scala/reflect/internal/PrintersTest.scala:730]: Its original is null (I hope so :)). I think wasEmpty is false (I can't check it right now, I'll see a bit later). Since (wasEmpty || isEmpty) is false, TypeTree is not empty, so its tpe is used for printing.

File: src/reflect/scala/reflect/internal/Trees.scala (hunk 2 above)
Comment (VladimirNik) [inline src/reflect/scala/reflect/internal/Trees.scala]: Thank you! I thought that var was private[this]

File: src/reflect/scala/reflect/internal/Printers.scala (hunk 5 above)
Comment (VladimirNik) [inline src/reflect/scala/reflect/internal/Printers.scala]: Fixed

File: src/reflect/scala/reflect/internal/Printers.scala (hunk 3 above)
Comment (VladimirNik) [inline src/reflect/scala/reflect/internal/Printers.scala:1076]: I don't think so, if tree has original we should print original. Based on wasEmpty we can't say if tree contains original or not. For example:

``` scala
q"val a: ${typeOf[Int]} = ..." //wasEmpty == false, original == null - for Int
toolbox.typecheck(q"val a: Int = ...") //wasEmpty == false for Int, original != null - for Int
```

File: test/junit/scala/reflect/internal/PrintersTest.scala (hunk 1 above)
Comment (VladimirNik) [inline test/junit/scala/reflect/internal/PrintersTest.scala:730]: Yes, wasEmpty is false

Comment (xeno-by) [inline test/junit/scala/reflect/internal/PrintersTest.scala:730]: Crazy :)

File: src/reflect/scala/reflect/internal/Printers.scala (hunk 4 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Printers.scala:997]: @densh could you elaborate?

**Comment by retronym** [comment]
Can you please resubmit this against the 2.11.x branch, and finish the review process on the new PR? Thanks!

**Comment by VladimirNik** [comment]
Resubmitted [pull request](https://github.com/scala/scala/pull/3736)
